visualization_scripts/run.py

```diff
@@ -0,0 +1,175 @@
+
+    # check in last line for process name
+    last_line = output.strip().split(b'\r\n')[-1]
+
+    return last_line.lower().startswith(bytes_name.lower())
+
+
+def find_path(name, path):
+    """
+    Search Algorithm to find Cytoscape application on machine
+
+    Args:
+        name (str): name of application 
+        path (str): top directory to start search from 
+
+    Returns:
+        path to first match on machine
+    """
+
+    # list to hold found matches
+    result = []
+
+    # walk directory/file tree 
+    for root, dirs, files in os.walk(path):
+        if name in files:
+            result.append(os.path.join(root, name))
+    return result[0]
+
+
+def vis(output, style):
+    """
+    REQUIRED ARGS:
+        output: output file from tps software
+        style: style file for cytoscpe 
+    """
+    
+    #create cyrest client
+    cyy = cyrest.cyclient()
+    
+    # Step 1: Create py2cytoscape client
+    cy = CyRestClient()
+
+    # Reset
+    cy.session.delete()
+    print("---session reset")
+
+    # Step 2: Load network from somewhere
+    print("---loading network file")
+    tps_net = cy.network.create_from(output)
+    print("---Done")
+    time.sleep(2)
+
+
+    # Step 3: Apply layout
+    # list of styles to apply
+    # only contains the one style loaded from style file 
+    print("---apply style file")
+    style = cyy.vizmap.load_file(style)
+    print("---style name: ", style)
+    cyy.vizmap.apply(style[0])
+
+    print("Done! open up cytoscape session")
+
+
+
+def main(args):
+
+    # input args from command line
+    OUTPUT_FILE = args[1]
+    STYLE_FILE = args[2]
+    DIRNAME = r"C:\Users\ajshe\OneDrive\Documents\Comp_bio\Cytoscape_v3.7.1"
+    CYTOSCAPE = 'Cytoscape.exe'
+
+    # check if Cytoscape is running 
+    isRunning = process_exists('Cytoscape.exe')
+
+    if (isRunning == False):
+        print("---Cytoscape not running")
+
+        # find path to Cytoscape on machine 
+        cytoPath = find_path(CYTOSCAPE, DIRNAME)
+
+        # open Cytoscape 
+        p = subprocess.Popen(cytoPath)
+
+    else:
+
+        # continue on
+        print("---Cytoscape running")
+
+    # call vis fuction to load input files in Cytoscape session
+    # catch errors: ConnectionError, JSONDecoderError
+    connectionCount = 0 
+    JSONCount = 0
+    switch = False
+    while switch == False:
+        try:
+
+            # run visualization function 
+            vis(OUTPUT_FILE, STYLE_FILE)
+
+            # print information --------------------------------
+            print("---CyRest client created")
+            print("---output file: " + OUTPUT_FILE)
+            print("---style file: " + STYLE_FILE)
+            print("---ConnectionError catches: ", connectionCount)
+            print("---JSONDecoderError catches: ", JSONCount)
+            # --------------------------------------------------
+
+            # flip switch value
+            switch = True
+
+        except CE as e:

```

Comment:
Did we want to move the sleep command to inside these `except` blocks?

Reply:
Like I mentioned last time we met, these errors are being thrown between 60-100 times during the execution of the script. It would be most efficient to sleep the program in order to not waste CPU cycles. After running a couple of tests, the program takes around 30 seconds to load the Cytoscape session with no sleep, but 100 or so errors are thrown in the process. Whereas if I sleep the program after a ConnectionError is thrown for 15 seconds, it takes around 40 seconds to complete, but only 2 errors are thrown throughout execution. We can work to try to optimize this, but sleeping the program definitely reduced the amount of errors being thrown (as expected).